### PR TITLE
Don't block an app's main thread when sending typed characters to NVDA

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Re #11369 , #11398  

### Summary of the issue:
Currently, the main way that NVDA reports characters being typed is by detecting them in-process by watching for WM_CHAR window messages, and then sending the characters to NVDA via an RPC call.
However, the rpc call to NvDA is made directly from the app's main thread where the message was detected, and therefore blocks that thread until the rpc call is fully executed.
This can cause a deadlock between multiple threads in NVDA, bit more importantly a deadlock between NVDA and the app in question, if a cross-process COM object for the app happens to be released by Python's garbage collector within the incoming RPC call to NVDA, as this COM object's Release method will try and execute via the Windows message loop on the app's main thread, which is currently blocked by the rpc call.
  
### Description of how this pull request fixes the issue:
Rather than making the RPC call directly, queue it to NVDA's in-process manager thread using APC (Asynchronous Procedure Calls) , and execute it there, no longer blocking the app's main thread.
 
### Testing performed:
* Ran NVDA with speak typed characters turned on. Ensured that typed characters could be heard in various  situations such as in Microsoft Word, Notepad, and the Run dialog.
* Placed temporarily logging in the in-process code, to verify that the rpc call was in deed being made from the manager thread.

### Known issues with pull request:
Not an issue, but just a note for future: In Windows 10 1607 and higher, we are able to detect typed characters within NVDA itself, rather than having to rely on in-process code at all. We should consider disabling the typed character support in-process completely on these more recent versions of Windows 10. However, right now it is more important to address the issue at its source, where it affects all Operating System versions.

### Change log entry:
Bug fixes:
- NVDA should no longer freeze in Microsoft word when rapidly typing characters.

### CODE REVIEW

**Comment by leonardder** [comment]
Just out of curiosity, have you ever considered moving to async RPC to avoid issues like this?

File: nvdaHelper/remote/typedCharacter.cpp [hunk 1]
```diff
@@ -21,14 +21,21 @@ This license can be found at:
...
+
 LRESULT CALLBACK typedCharacter_getMessageHook(int code, WPARAM wParam, LPARAM lParam) {
 	static WPARAM lastCharacter=0;
 	MSG* pmsg=(MSG*)lParam;
 	if(pmsg->message==WM_KEYDOWN) {
 		typedCharacter_window=pmsg->hwnd;
 		lastCharacter=0;
 	} else if((typedCharacter_window!=0)&&(pmsg->message==WM_CHAR)&&(pmsg->hwnd==typedCharacter_window)&&(pmsg->wParam!=lastCharacter)) { 
-		nvdaControllerInternal_typedCharacterNotify(GetCurrentThreadId(),static_cast<wchar_t>(pmsg->wParam));
+		// Instruct NVDA's inproc manager thread to report the typed chracter to NVDA via rpc.
```
Comment (leonardder) [inline nvdaHelper/remote/typedCharacter.cpp:35]: Typo

```suggestion
		// Instruct NVDA's inproc manager thread to report the typed character to NVDA via rpc.

```

**Review by LeonarddeR — APPROVED** [review]

**Comment by michaelDCurran** [comment]
I looked into async rpc, but it really is quite complex. On the client 
side, for every call you need to initialize an async handle, and have 
code that can handle the responce. On the server side, you need to call 
a completion function at the end, and I think arguments are handled 
differently. what we really need is simply just an rpc function that 
returns straight away and doesn't care about handling a return value.


We currently use the ncalrpc protocol, which is certainly the fastest 
protocol as a lot of it is at kernel level and it is written to be 
purely local. The one draw back is that is not cancellable. If we used a 
network protocol we could cancel, or we could use attributes in the idl 
like message, or maybe, or idenpotent, but I think this type of protocl 
may be slower. But most importantly: I'm not sure how easy it would be 
to get the security right. We had to do a lot of complex work to support 
Metro apps in windows 8.

Thus for now at least, I think queuing some of our rpc calls to a 
non-main thread is the best way to go.

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Some small changes requested, but overall looks good!

**Review by feerrenrut — APPROVED** [review]
Thanks @michaelDCurran

**Pull request merged**